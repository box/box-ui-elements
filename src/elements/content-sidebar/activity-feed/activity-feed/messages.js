/**
 * @flow
 * @file i18n messages
 * @author Box
 */

import { defineMessages } from 'react-intl';

const messages = defineMessages({
    annotationMissingError: {
        id: 'be.activitySidebar.activityFeed.annotationMissingError',
        defaultMessage: 'This comment no longer exists',
        description: 'Text to show when an annotation activity no longer exists',
    },
    feedInlineErrorTitle: {
        id: 'be.activitySidebar.activityFeed.feedInlineErrorTitle',
        defaultMessage: 'Error',
        description: 'Error title',
    },
    taskMissingError: {
        id: 'be.activitySidebar.activityFeed.taskMissingError',
        defaultMessage: 'This task no longer exists',
        description: 'Text to show when a task no longer exists',
    },
    commentMissingError: {
        id: 'be.activitySidebar.activityFeed.commentMissingError',
        defaultMessage: 'This comment no longer exists',
        description: 'Text to show when comment no longer exists',
    },
});

export default messages;
