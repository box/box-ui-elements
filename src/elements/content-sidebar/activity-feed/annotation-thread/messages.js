import { defineMessages } from 'react-intl';

const messages = defineMessages({
    errorFetchAnnotation: {
        id: 'be.annotattionThread.errorFetchAnnotation',
        description: 'Error message when an annotation fetch fails',
        defaultMessage: 'The annotation could not be fetched.',
    },
    errorEditAnnotation: {
        id: 'be.annotationThread.errorEditAnnotation',
        description: 'Error message when an annotation update fails',
        defaultMessage: 'This annotation could not be modified.',
    },
    errorDeleteAnnotation: {
        id: 'be.annotationThread.errorDeleteAnnotation',
        description: 'Error message when an annotation deletion fails',
        defaultMessage: 'There was an error deleting this item.',
    },
});

export default messages;
