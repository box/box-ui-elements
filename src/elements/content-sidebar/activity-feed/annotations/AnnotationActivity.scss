@import '../../../common/variables';

.bcs-AnnotationActivity {
    position: relative;
    z-index: 1;
}

// the comment activity item
.bcs-AnnotationActivity-media {
    opacity: 1;
    transition: opacity .33s;

    &.bcs-is-pending {
        opacity: .4;
        pointer-events: none;
    }
}

.bcs-AnnotationActivity-headline {
    // Because the annotation menu is absolutely positioned,
    // this padding reserves that space so long headlines won't overlap with the menu
    margin-right: 26px;
    // target only Edge because this breaks other browsers
    // but fixes Edge
    @supports (-ms-ime-align: auto) {
        word-break: break-all;
    }
}

// tethered styles for delete confirmation modal
.bcs-AnnotationActivity-deleteConfirmationModal {
    z-index: $overlay-z-index;
}

.bcs-AnnotationActivity-menu {
    position: absolute;
    top: $sidebarActivityFeedSpacingVertical;
    right: $sidebarActivityFeedSpacingHorizontal;
    z-index: 1;
}

.bcs-AnnotationActivity-timestamp {
    display: flex;
    white-space: nowrap;

    .bcs-ActivityTimestamp {
        &::after {
            margin-right: 8px;
            margin-left: 8px;
            color: $bdl-gray-50;
            content: '\2022';
        }
    }
}

// override for activity feed defaults
.be {
    .bcs-AnnotationActivity-editor.bcs-CommentForm {
        margin-top: 20px;
        padding: 0;
        border-top: none;
    }

    .bcs-AnnotationActivity-link {
        display: flex;
        font-weight: bold;
        font-size: 12px;
    }
}
