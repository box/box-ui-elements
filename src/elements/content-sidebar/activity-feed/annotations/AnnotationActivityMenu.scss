@import '../../../common/variables';

.bcs-AnnotationActivityMenu-unresolveAnnotation path {
    fill: $bdl-gray;
}
