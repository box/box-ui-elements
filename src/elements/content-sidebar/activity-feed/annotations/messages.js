import { defineMessages } from 'react-intl';

const messages = defineMessages({
    annotationActivityDeleteMenuItem: {
        id: 'be.contentSidebar.activityFeed.annotationActivity.annotationActivityDeleteMenuItem',
        defaultMessage: 'Delete',
        description: 'Text to show on menu item to delete the annotation comment',
    },
    annotationActivityDeletePrompt: {
        id: 'be.contentSidebar.activityFeed.annotationActivity.annotationActivityDeletePrompt',
        defaultMessage: 'Are you sure you want to permanently delete this comment?',
        description: 'Confirmation prompt text to delete the annotation comment',
    },
    annotationActivityEditMenuItem: {
        id: 'be.contentSidebar.activityFeed.annotationActivity.annotationActivityEditMenuItem',
        defaultMessage: 'Modify',
        description: 'Text to show on menu item to edit the annotation comment',
    },
    annotationActivityPostedFullDateTime: {
        id: 'be.contentSidebar.activityFeed.annotationActivity.annotationActivityPostedFullDateTime',
        defaultMessage: '{time, date, full} at {time, time, short}',
        description: 'Full data and time for annotation title',
    },
    annotationActivityPageItem: {
        id: 'be.contentSidebar.activityFeed.annotationActivityPageItem',
        defaultMessage: 'Page {number}',
        description: 'Annotation activity item link shown on annotation activity',
    },
    annotationActivityVersionLink: {
        id: 'be.contentSidebar.activityFeed.annotationActivityVersionLink',
        defaultMessage: 'Version {number}',
        description: 'Annotation activity item link shown on annotation activity for previous file version',
    },
    annotationActivityVersionUnavailable: {
        id: 'be.contentSidebar.activityFeed.annotationActivityVersionUnavailable',
        defaultMessage: 'Version Unavailable',
        description:
            'Annotation activity item link shown on annotation activity for previous file version that is unavailable',
    },
});

export default messages;
