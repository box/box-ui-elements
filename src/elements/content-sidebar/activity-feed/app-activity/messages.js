/**
 * @flow
 * @file i18n messages
 * @author Box
 */

import { defineMessages } from 'react-intl';

const messages = defineMessages({
    appActivityDeleteMenuItem: {
        id: 'be.appActivity.deleteMenuItem',
        defaultMessage: 'Delete',
        description: 'Text to display in the delete app activity submit button',
    },
    appActivityDeletePrompt: {
        id: 'be.appActivity.deletePrompt',
        defaultMessage: 'Are you sure you want to permanently delete this app activity?',
        description: 'Confirmation prompt text to delete app activity',
    },
    appActivityAltIcon: {
        id: 'be.appActivity.altIcon',
        defaultMessage: '{appActivityName} Icon',
        description: 'Alt message if app activity icon is missing or cannot load',
    },
});

export default messages;
