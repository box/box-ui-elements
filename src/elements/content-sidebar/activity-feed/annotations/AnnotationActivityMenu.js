// @flow
import * as React from 'react';
import classNames from 'classnames';
import { FormattedMessage } from 'react-intl';
import DropdownMenu from '../../../../components/dropdown-menu';
import IconEllipsis from '../../../../icons/general/IconEllipsis';
import messages from './messages';
import Pencil16 from '../../../../icon/line/Pencil16';
import PlainButton from '../../../../components/plain-button';
import Trash16 from '../../../../icon/fill/Trash16';
import { ACTIVITY_TARGETS } from '../../../common/interactionTargets';
import { bdlGray50 } from '../../../../styles/variables';
import { Menu, MenuItem } from '../../../../components/menu';

type AnnotationActivityMenuProps = {
    canDelete?: boolean,
    canEdit?: boolean,
    className?: string,
    id: string,
    isDisabled?: boolean,
    onDelete: () => void,
    onEdit: () => void,
    onMenuClose: () => void,
    onMenuOpen: () => void,
};

const AnnotationActivityMenu = ({
    canDelete,
    canEdit,
    className,
    id,
    isDisabled,
    onDelete,
    onEdit,
    onMenuClose,
    onMenuOpen,
}: AnnotationActivityMenuProps) => {
    const menuProps = {
        'data-resin-component': 'preview',
        'data-resin-feature': 'annotations',
    };

    return (
        <DropdownMenu constrainToScrollParent isRightAligned onMenuClose={onMenuClose} onMenuOpen={onMenuOpen}>
            <PlainButton
                className={classNames('bcs-AnnotationActivityMenu', className)}
                isDisabled={isDisabled}
                data-testid="annotation-activity-actions-menu"
                type="button"
            >
                <IconEllipsis color={bdlGray50} height={16} width={16} />
            </PlainButton>
            <Menu {...menuProps}>
                {canEdit && (
                    <MenuItem
                        data-resin-itemid={id}
                        data-resin-target={ACTIVITY_TARGETS.ANNOTATION_OPTIONS_EDIT}
                        data-testid="edit-annotation-activity"
                        onClick={onEdit}
                    >
                        <Pencil16 />
                        <FormattedMessage {...messages.annotationActivityEditMenuItem} />
                    </MenuItem>
                )}
                {canDelete && (
                    <MenuItem
                        data-resin-itemid={id}
                        data-resin-target={ACTIVITY_TARGETS.ANNOTATION_OPTIONS_DELETE}
                        data-testid="delete-annotation-activity"
                        onClick={onDelete}
                    >
                        <Trash16 />
                        <FormattedMessage {...messages.annotationActivityDeleteMenuItem} />
                    </MenuItem>
                )}
            </Menu>
        </DropdownMenu>
    );
};

export default AnnotationActivityMenu;
