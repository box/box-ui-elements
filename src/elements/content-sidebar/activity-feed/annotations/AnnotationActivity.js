// @flow
import * as React from 'react';
import classNames from 'classnames';
import getProp from 'lodash/get';
import noop from 'lodash/noop';
import TetherComponent from 'react-tether';
import ActivityError from '../common/activity-error';
import ActivityMessage from '../common/activity-message';
import ActivityTimestamp from '../common/activity-timestamp';
import AnnotationActivityLink from './AnnotationActivityLink';
import AnnotationActivityMenu from './AnnotationActivityMenu';
import Avatar from '../Avatar';
import CommentForm from '../comment-form/CommentForm';
import DeleteConfirmation from '../common/delete-confirmation';
import Media from '../../../../components/media';
import messages from './messages';
import SelectableActivityCard from '../SelectableActivityCard';
import UserLink from '../common/user-link';
import { ACTIVITY_TARGETS } from '../../../common/interactionTargets';
import { PLACEHOLDER_USER } from '../../../../constants';
import type { Annotation, AnnotationPermission } from '../../../../common/types/feed';
import type { GetAvatarUrlCallback, GetProfileUrlCallback } from '../../../common/flowTypes';
import type { SelectorItems, User } from '../../../../common/types/core';

import './AnnotationActivity.scss';

type Props = {
    currentUser?: User,
    getAvatarUrl: GetAvatarUrlCallback,
    getMentionWithQuery?: (searchStr: string) => void,
    getUserProfileUrl?: GetProfileUrlCallback,
    isCurrentVersion: boolean,
    item: Annotation,
    mentionSelectorContacts?: SelectorItems<User>,
    onDelete?: ({ id: string, permissions: AnnotationPermission }) => any,
    onEdit?: (id: string, text: string, permissions: AnnotationPermission) => void,
    onSelect?: (annotation: Annotation) => any,
};

const AnnotationActivity = ({
    currentUser,
    item,
    getAvatarUrl,
    getMentionWithQuery,
    getUserProfileUrl,
    isCurrentVersion,
    mentionSelectorContacts,
    onDelete = noop,
    onEdit = noop,
    onSelect = noop,
}: Props) => {
    const [isConfirmingDelete, setIsConfirmingDelete] = React.useState(false);
    const [isEditing, setIsEditing] = React.useState(false);
    const [isMenuOpen, setIsMenuOpen] = React.useState(false);
    const { created_at, created_by, description, error, file_version, id, isPending, permissions = {}, target } = item;
    const { can_delete: canDelete, can_edit: canEdit } = permissions;
    const isFileVersionUnavailable = file_version === null;
    const isCardDisabled = !!error || isConfirmingDelete || isMenuOpen || isEditing || isFileVersionUnavailable;
    const isMenuVisible = (canDelete || canEdit) && !isPending;

    const handleDelete = (): void => setIsConfirmingDelete(true);
    const handleDeleteCancel = (): void => setIsConfirmingDelete(false);
    const handleDeleteConfirm = (): void => {
        setIsConfirmingDelete(false);
        onDelete({ id, permissions });
    };
    const handleEdit = (): void => setIsEditing(true);
    const handleFormCancel = (): void => setIsEditing(false);
    const handleFormSubmit = ({ text }): void => {
        setIsEditing(false);
        onEdit(id, text, permissions);
    };
    const handleMenuClose = (): void => setIsMenuOpen(false);
    const handleMenuOpen = (): void => setIsMenuOpen(true);
    const handleMouseDown = (event: MouseEvent) => {
        if (isCardDisabled) {
            return;
        }

        // Prevents document event handlers from executing because box-annotations relies on
        // detecting mouse events on the document outside of annotation targets to determine when to
        // deselect annotations
        event.stopPropagation();
    };
    const handleSelect = () => onSelect(item);

    const createdAtTimestamp = new Date(created_at).getTime();
    const createdByUser = created_by || PLACEHOLDER_USER;
    const linkMessage = isCurrentVersion ? messages.annotationActivityPageItem : messages.annotationActivityVersionLink;
    const linkValue = isCurrentVersion ? target.location.value : getProp(file_version, 'version_number');
    const message = (description && description.message) || '';
    const activityLinkMessage = isFileVersionUnavailable
        ? messages.annotationActivityVersionUnavailable
        : { ...linkMessage, values: { number: linkValue } };
    const tetherProps = {
        attachment: 'top right',
        className: 'bcs-AnnotationActivity-deleteConfirmationModal',
        constraints: [{ to: 'scrollParent', attachment: 'together' }],
        targetAttachment: 'bottom right',
    };

    return (
        <>
            <SelectableActivityCard
                className="bcs-AnnotationActivity"
                data-resin-feature="annotations"
                data-resin-iscurrent={isCurrentVersion}
                data-resin-itemid={id}
                data-resin-target="annotationButton"
                isDisabled={isCardDisabled}
                onMouseDown={handleMouseDown}
                onSelect={handleSelect}
            >
                <Media
                    className={classNames('bcs-AnnotationActivity-media', {
                        'bcs-is-pending': isPending || error,
                    })}
                >
                    <Media.Figure>
                        <Avatar getAvatarUrl={getAvatarUrl} user={createdByUser} />
                    </Media.Figure>
                    <Media.Body>
                        <div className="bcs-AnnotationActivity-headline">
                            <UserLink
                                data-resin-target={ACTIVITY_TARGETS.PROFILE}
                                getUserProfileUrl={getUserProfileUrl}
                                id={createdByUser.id}
                                name={createdByUser.name}
                            />
                        </div>
                        <div className="bcs-AnnotationActivity-timestamp">
                            <ActivityTimestamp date={createdAtTimestamp} />
                            <AnnotationActivityLink
                                className="bcs-AnnotationActivity-link"
                                data-resin-target="annotationLink"
                                id={id}
                                isDisabled={isFileVersionUnavailable}
                                message={activityLinkMessage}
                                onClick={handleSelect}
                            />
                        </div>
                        {isEditing && currentUser ? (
                            <CommentForm
                                className="bcs-AnnotationActivity-editor"
                                entityId={id}
                                getAvatarUrl={getAvatarUrl}
                                getMentionWithQuery={getMentionWithQuery}
                                isEditing={isEditing}
                                isOpen={isEditing}
                                mentionSelectorContacts={mentionSelectorContacts}
                                onCancel={handleFormCancel}
                                updateComment={handleFormSubmit}
                                user={currentUser}
                                tagged_message={message}
                            />
                        ) : (
                            <ActivityMessage id={id} tagged_message={message} getUserProfileUrl={getUserProfileUrl} />
                        )}
                    </Media.Body>
                </Media>
                {/* $FlowFixMe */}
                {error ? <ActivityError {...error} /> : null}
            </SelectableActivityCard>
            <TetherComponent {...tetherProps}>
                {isMenuVisible && (
                    <AnnotationActivityMenu
                        canDelete={canDelete}
                        canEdit={canEdit}
                        className="bcs-AnnotationActivity-menu"
                        id={id}
                        isDisabled={isConfirmingDelete}
                        onDelete={handleDelete}
                        onEdit={handleEdit}
                        onMenuClose={handleMenuClose}
                        onMenuOpen={handleMenuOpen}
                    />
                )}
                {isConfirmingDelete && (
                    <DeleteConfirmation
                        data-resin-component={ACTIVITY_TARGETS.ANNOTATION_OPTIONS}
                        isOpen={isConfirmingDelete}
                        message={messages.annotationActivityDeletePrompt}
                        onDeleteCancel={handleDeleteCancel}
                        onDeleteConfirm={handleDeleteConfirm}
                    />
                )}
            </TetherComponent>
        </>
    );
};

export default AnnotationActivity;
