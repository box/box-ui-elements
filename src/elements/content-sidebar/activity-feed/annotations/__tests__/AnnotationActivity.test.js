import * as React from 'react';
import { shallow } from 'enzyme';

import AnnotationActivity from '../AnnotationActivity';
import AnnotationActivityMenu from '../AnnotationActivityMenu';
import CommentForm from '../../comment-form/CommentForm';
import DeleteConfirmation from '../../common/delete-confirmation';
import Media from '../../../../../components/media';
import messages from '../messages';
import SelectableActivityCard from '../../SelectableActivityCard';

jest.mock('../../Avatar', () => () => 'Avatar');

const currentUser = {
    name: 'testuser',
    id: 11,
};
const mentionSelectorContacts = [];
const TIME_STRING_SEPT_27_2017 = '2017-09-27T10:40:41-07:00';

const allHandlers = {
    contacts: {
        getMentionWithQuery: jest.fn(),
    },
};

describe('elements/content-sidebar/ActivityFeed/annotations/AnnotationActivity', () => {
    const mockAnnotation = {
        created_at: TIME_STRING_SEPT_27_2017,
        created_by: { name: 'Jane Doe', id: 10 },
        description: { message: 'test' },
        file_version: {
            id: '456',
            version_number: '2',
        },
        id: '123',
        target: { location: { value: 1 } },
    };
    const mockActivity = {
        currentUser,
        handlers: allHandlers,
        isCurrentVersion: true,
        item: mockAnnotation,
        mentionSelectorContacts,
    };

    const getWrapper = (props = {}) => shallow(<AnnotationActivity {...mockActivity} {...props} />);

    beforeEach(() => {
        CommentForm.default = jest.fn().mockReturnValue(<div />);
    });

    test('should not render annotation activity menu when both can_delete is false and can_edit is false', () => {
        const item = {
            ...mockAnnotation,
            permissions: { can_delete: false, can_edit: false },
        };

        const wrapper = getWrapper({ item });

        expect(wrapper.exists(AnnotationActivityMenu)).toBe(false);
    });

    test.each`
        canDelete | canEdit
        ${false}  | ${true}
        ${true}   | ${false}
        ${true}   | ${true}
    `(
        'should correctly render annotation activity when canDelete: $canDelete and canEdit: $canEdit',
        ({ canDelete, canEdit }) => {
            const unixTime = new Date(TIME_STRING_SEPT_27_2017).getTime();
            const item = {
                ...mockAnnotation,
                permissions: { can_delete: canDelete, can_edit: canEdit },
            };

            const wrapper = getWrapper({ item });

            expect(wrapper.find('ActivityTimestamp').prop('date')).toEqual(unixTime);
            expect(wrapper.find('AnnotationActivityLink').props()).toMatchObject({
                'data-resin-target': 'annotationLink',
                message: {
                    ...messages.annotationActivityPageItem,
                    values: { number: 1 },
                },
            });
            expect(wrapper.exists(AnnotationActivityMenu)).toBe(true);
            expect(wrapper.find('ActivityMessage').prop('tagged_message')).toEqual(
                mockActivity.item.description.message,
            );
        },
    );

    test('should render CommentForm if user clicks on the Modify menu item', () => {
        const activity = {
            item: {
                ...mockAnnotation,
                isPending: false,
                permissions: { can_edit: true },
            },
        };

        const wrapper = getWrapper({ ...mockActivity, ...activity });

        wrapper
            .find(AnnotationActivityMenu)
            .dive()
            .simulate('click');
        wrapper
            .find(AnnotationActivityMenu)
            .dive()
            .find('MenuItem')
            .simulate('click');
        expect(wrapper.exists('CommentForm')).toBe(true);

        // Firing the onCancel prop will remove the CommentForm
        wrapper
            .find('CommentForm')
            .dive()
            .props()
            .onCancel();
        expect(wrapper.exists('CommentForm')).toBe(false);
    });

    test('should correctly render annotation activity of another file version', () => {
        const wrapper = getWrapper({ isCurrentVersion: false });

        expect(wrapper.find('AnnotationActivityLink').prop('message')).toEqual({
            ...messages.annotationActivityVersionLink,
            values: { number: '2' },
        });
    });

    test('should render version unavailable if file version is null', () => {
        const wrapper = getWrapper({ item: { ...mockAnnotation, file_version: null } });
        const activityLink = wrapper.find('AnnotationActivityLink');

        expect(activityLink.prop('message')).toEqual({
            ...messages.annotationActivityVersionUnavailable,
        });
        expect(activityLink.prop('isDisabled')).toBe(true);
    });

    test('should render commenter as a link', () => {
        const wrapper = getWrapper();
        expect(wrapper.find('UserLink').prop('name')).toEqual(mockActivity.item.created_by.name);
    });

    test('should not show actions menu when annotation activity is pending', () => {
        const item = {
            ...mockAnnotation,
            permissions: { can_delete: true },
            isPending: true,
        };

        const wrapper = getWrapper({ item });

        expect(wrapper.exists(AnnotationActivityMenu)).toBe(false);
        expect(wrapper.find(Media).hasClass('bcs-is-pending')).toBe(true);
    });

    test('should render an error when one is defined', () => {
        const activity = {
            item: {
                ...mockAnnotation,
                error: {
                    title: 'error',
                    message: 'This is an error message',
                },
            },
            onDelete: jest.fn(),
        };

        const wrapper = getWrapper(activity);
        expect(wrapper.find('ActivityError').length).toEqual(1);
    });

    test('should render an error cta when an action is defined', () => {
        const onActionSpy = jest.fn();
        const activity = {
            item: {
                ...mockAnnotation,
                error: {
                    title: 'error',
                    message: 'This is an error message',
                    action: {
                        text: 'click',
                        onAction: onActionSpy,
                    },
                },
            },
            onDelete: jest.fn(),
        };

        const wrapper = mount(<AnnotationActivity {...mockActivity} {...activity} />);

        const inlineErrorActionLink = wrapper.find('InlineError').find('button.bcs-ActivityError-action');
        expect(inlineErrorActionLink.length).toEqual(1);

        const inlineErrorActionOnClick = inlineErrorActionLink.prop('onClick');

        inlineErrorActionOnClick();

        expect(onActionSpy).toHaveBeenCalledTimes(1);
    });

    describe('delete confirmation behavior', () => {
        test('should render the DeleteConfirmation when delete menu item is selected', () => {
            const item = {
                ...mockAnnotation,
                permissions: { can_delete: true },
            };

            const wrapper = getWrapper({ item });

            wrapper.find(AnnotationActivityMenu).prop('onDelete')();

            expect(wrapper.exists(DeleteConfirmation));
        });

        test('should close the DeleteConfirmation when cancel is selected', () => {
            const item = {
                ...mockAnnotation,
                permissions: { can_delete: true },
            };

            const wrapper = getWrapper({ item });

            wrapper.find(AnnotationActivityMenu).prop('onDelete')();
            wrapper.find(DeleteConfirmation).prop('onDeleteCancel')();

            expect(wrapper.exists(DeleteConfirmation)).toBe(false);
        });

        test('should call onDelete when confirm delete is selected', () => {
            const onDelete = jest.fn();
            const permissions = { can_delete: true };
            const item = {
                ...mockAnnotation,
                permissions,
            };

            const wrapper = getWrapper({ item, onDelete });

            wrapper.find(AnnotationActivityMenu).prop('onDelete')();
            wrapper.find(DeleteConfirmation).prop('onDeleteConfirm')();

            expect(onDelete).toHaveBeenCalledWith({ id: mockAnnotation.id, permissions });
            expect(wrapper.exists(DeleteConfirmation)).toBe(false);
        });
    });

    describe('SelectableActivityCard', () => {
        const getActivityItem = overrides => ({
            ...mockAnnotation,
            permissions: { can_delete: true, can_edit: true },
            ...overrides,
        });

        test('should render as SelectableActivityCard', () => {
            const wrapper = getWrapper();

            expect(wrapper.exists(SelectableActivityCard)).toBe(true);
            expect(wrapper.find(SelectableActivityCard).props()).toMatchObject({
                className: 'bcs-AnnotationActivity',
                'data-resin-iscurrent': true,
                'data-resin-itemid': mockAnnotation.id,
                'data-resin-feature': 'annotations',
                'data-resin-target': 'annotationButton',
                isDisabled: false,
                onMouseDown: expect.any(Function),
                onSelect: expect.any(Function),
            });
        });

        test('should disable card if there is an error', () => {
            const activity = {
                item: {
                    ...mockAnnotation,
                    error: {
                        title: 'error',
                        message: 'This is an error message',
                        action: {
                            text: 'click',
                        },
                    },
                },
            };
            const wrapper = getWrapper(activity);

            expect(wrapper.find(SelectableActivityCard).prop('isDisabled')).toBe(true);
        });

        test('should disable card if the overflow menu is open', () => {
            const wrapper = getWrapper({ item: getActivityItem() });

            expect(wrapper.find(SelectableActivityCard).prop('isDisabled')).toBe(false);

            wrapper.find(AnnotationActivityMenu).prop('onMenuOpen')();

            expect(wrapper.find(SelectableActivityCard).prop('isDisabled')).toBe(true);
        });

        test('should disable card if editing the comment', () => {
            const wrapper = getWrapper({ item: getActivityItem() });

            expect(wrapper.find(SelectableActivityCard).prop('isDisabled')).toBe(false);

            wrapper.find(AnnotationActivityMenu).prop('onEdit')();

            expect(wrapper.find(SelectableActivityCard).prop('isDisabled')).toBe(true);
        });

        test('should disable card if file version is unavailable', () => {
            const wrapper = getWrapper({ item: getActivityItem({ file_version: null }) });

            expect(wrapper.find(SelectableActivityCard).prop('isDisabled')).toBe(true);
        });

        test('should disable card if the delete confirmation is open', () => {
            const wrapper = getWrapper({ item: getActivityItem() });

            expect(wrapper.find(SelectableActivityCard).prop('isDisabled')).toBe(false);

            wrapper.find(AnnotationActivityMenu).prop('onDelete')();

            expect(wrapper.find(SelectableActivityCard).prop('isDisabled')).toBe(true);
        });

        test('should stop propagation of mousedown event from SelectableActivityCard', () => {
            const event = { stopPropagation: jest.fn() };
            const wrapper = getWrapper({ item: getActivityItem() });

            wrapper.find(SelectableActivityCard).simulate('mousedown', event);

            expect(event.stopPropagation).toHaveBeenCalled();
        });

        test('should not stop propagation of mousedown event from SelectableActivityCard when disabled', () => {
            const event = { stopPropagation: jest.fn() };
            const wrapper = getWrapper({ item: getActivityItem({ file_version: null }) });

            expect(wrapper.find(SelectableActivityCard).prop('isDisabled')).toBe(true);

            wrapper.find(SelectableActivityCard).simulate('mousedown', event);

            expect(event.stopPropagation).not.toHaveBeenCalled();
        });
    });
});
