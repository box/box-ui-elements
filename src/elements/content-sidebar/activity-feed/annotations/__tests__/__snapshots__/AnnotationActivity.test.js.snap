// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`elements/content-sidebar/ActivityFeed/annotations/AnnotationActivity should correctly render annotation activity 1`] = `
<div
  className="bcs-AnnotationActivity "
>
  <Media
    as="div"
    className="bcs-AnnotationActivity-media"
  >
    <MediaFigure
      as="figure"
    >
      <Component
        user={
          Object {
            "id": 10,
            "name": "50 Cent",
          }
        }
      />
    </MediaFigure>
    <MediaBody>
      <TetherComponent
        attachment="top right"
        className="bcs-AnnotationActivity-deleteConfirmationModal"
        constraints={
          Array [
            Object {
              "attachment": "together",
              "to": "scrollParent",
            },
          ]
        }
        renderElementTag="div"
        renderElementTo={null}
        targetAttachment="bottom right"
      >
        <MediaMenu
          data-testid="annotation-activity-actions-menu"
          dropdownProps={Object {}}
          isDisabled={false}
          menuProps={
            Object {
              "data-resin-component": "activityfeed-annotationactivity",
            }
          }
        >
          <MenuItem
            data-resin-target="activityfeed-annotationactivity-delete"
            data-testid="delete-annotation-activity"
            onClick={[Function]}
          >
            <IconTrash
              color="#222"
            />
            <FormattedMessage
              defaultMessage="Delete"
              id="be.contentSidebar.activityFeed.annotationActivity.annotationActivityDeleteMenuItem"
            />
          </MenuItem>
        </MediaMenu>
      </TetherComponent>
      <div
        className="bcs-AnnotationActivity-headline"
      >
        <UserLink
          data-resin-target="activityfeed-profilelink"
          id={10}
          name="50 Cent"
        />
      </div>
      <div>
        <ActivityTimestamp
          date={1506534041000}
        />
      </div>
      <ActivityMessage
        id="123"
        tagged_message="test"
        translationEnabled={false}
      />
      <AnnotationActivityLink
        id="123"
        message={
          Object {
            "defaultMessage": "Page {number}",
            "id": "be.contentSidebar.activityFeed.annotationActivityPageItem",
            "values": Object {
              "number": 1,
            },
          }
        }
        onClick={[Function]}
      />
    </MediaBody>
  </Media>
</div>
`;

exports[`elements/content-sidebar/ActivityFeed/annotations/AnnotationActivity should render an error when one is defined 1`] = `
<div
  className="bcs-AnnotationActivity "
>
  <Media
    as="div"
    className="bcs-AnnotationActivity-media bcs-is-pending"
  >
    <MediaFigure
      as="figure"
    >
      <Component
        user={
          Object {
            "id": 10,
            "name": "50 Cent",
          }
        }
      />
    </MediaFigure>
    <MediaBody>
      <div
        className="bcs-AnnotationActivity-headline"
      >
        <UserLink
          data-resin-target="activityfeed-profilelink"
          id={10}
          name="50 Cent"
        />
      </div>
      <div>
        <ActivityTimestamp
          date={1506534041000}
        />
      </div>
      <ActivityMessage
        id="123"
        tagged_message="test"
        translationEnabled={false}
      />
      <AnnotationActivityLink
        id="123"
        message={
          Object {
            "defaultMessage": "Page {number}",
            "id": "be.contentSidebar.activityFeed.annotationActivityPageItem",
            "values": Object {
              "number": 1,
            },
          }
        }
        onClick={[Function]}
      />
    </MediaBody>
  </Media>
  <ActivityError
    message="This is an error message"
    title="error"
  />
</div>
`;

exports[`elements/content-sidebar/ActivityFeed/annotations/AnnotationActivity should render commenter as a link 1`] = `
<div
  className="bcs-AnnotationActivity "
>
  <Media
    as="div"
    className="bcs-AnnotationActivity-media"
  >
    <MediaFigure
      as="figure"
    >
      <Component
        user={
          Object {
            "id": 10,
            "name": "50 Cent",
          }
        }
      />
    </MediaFigure>
    <MediaBody>
      <div
        className="bcs-AnnotationActivity-headline"
      >
        <UserLink
          data-resin-target="activityfeed-profilelink"
          id={10}
          name="50 Cent"
        />
      </div>
      <div>
        <ActivityTimestamp
          date={1506534041000}
        />
      </div>
      <ActivityMessage
        id="123"
        tagged_message="test"
        translationEnabled={false}
      />
      <AnnotationActivityLink
        id="123"
        message={
          Object {
            "defaultMessage": "Page {number}",
            "id": "be.contentSidebar.activityFeed.annotationActivityPageItem",
            "values": Object {
              "number": 1,
            },
          }
        }
        onClick={[Function]}
      />
    </MediaBody>
  </Media>
</div>
`;
